crates/containerd-shim-wasm/src/sandbox/cgroups.rs

```diff
@@ -0,0 +1,674 @@
+    if v1.join("cpu").exists() {
+        let stat = statfs(v1.join("cpu").as_path()).map_err(|e| std::io::Error::from(e))?;
+        let is_v1 = stat.filesystem_type() == nix::sys::statfs::CGROUP_SUPER_MAGIC;
+        if is_v1 {
+            if v2.exists() {
+                // Check if we are running in a hybrid cgroup setup
+                let data = std::fs::read(v2.join("cgroup.controllers"))
+                    .map_err(|e| std::io::Error::from(e))?;
+
+                let trimmed = std::str::from_utf8(&data)
+                    .map_err(|e| {
+                        Error::Others(format!(
+                            "could not convert cgroup.controllers to string: {}",
+                            e
+                        ))
+                    })?
+                    .trim();
+
+                if trimmed.len() > 0 {
+                    return Err(Error::FailedPrecondition(
+                        "hybyrid cgroup is not supported".to_string(),
+                    ));
+                }
+
+                return Ok(Box::new(CgroupV1 {
+                    base: v1,
+                    path: PathBuf::from(name),
+                }));
+            }
+            return Ok(Box::new(CgroupV1::new(v1, PathBuf::from(name))));
+        }
+    }
+
+    if v2.exists() {
+        return Ok(Box::new(CgroupV2::new(v2, PathBuf::from(name))));
+    }
+
+    if v1.exists() {
+        let stat = nix::sys::statfs::statfs(v1.as_path()).map_err(|e| std::io::Error::from(e))?;
+        if stat.filesystem_type() == nix::sys::statfs::CGROUP2_SUPER_MAGIC {
+            // cgroup2 is mounted directly on /sys/fs/cgroup
+            return Ok(Box::new(CgroupV2::new(v1, PathBuf::from(name))));
+        }
+    }
+
+    // slow path
+
+    let mut v1_found = false;
+    let mut v2_found = false;
+    let mut base = PathBuf::from("/");
+
+    // It's possible for V1 controllers to be mounted all over the place... this code does not support that.
+    // It is expected that all v1 controllers are mounted under the same path
+    for mount in MountIter::new()? {
+        let mount = mount?;
+        match mount.fstype.as_str() {
+            "cgroup" => {
+                base = mount.dest.parent().unwrap().to_path_buf();
+                v1_found = true;
+            }
+            "cgroup2" => {
+                base = mount.dest;
+                v2_found = true;
+            }
+            _ => {}
+        }
+        if v1_found && v2_found {
+            break;
+        }
+    }
+
+    if v1_found && v2_found {
+        let p = PathBuf::from(base.clone()).join("cgroup.controllers");
+        if std::fs::read(p).map_err(|e| std::io::Error::from(e))?.len() > 0 {
+            return Ok(Box::new(CgroupV1 {
+                base: base,
+                path: PathBuf::from(name),
+            }));
+        }
+        return Err(Error::FailedPrecondition(
+            "hybyrid cgroup is not supported".to_string(),
```

Comment:
Why is the logic here to detect hybyrid cgroup different from the fast path? Specifically, in the fast path, we converted `std::fs::read(v2.join("cgroup.controllers"))` to `trimmed` string and then checked length of the string. Here, we only checked `std::fs::read(p).len() > 0`.